subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java

```diff
@@ -0,0 +1,304 @@
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Optional;
+
+public class DefaultCachingStateBuilder implements CachingStateBuilder {
+    private ImplementationSnapshot implementation;
+    private final ImmutableList.Builder<ImplementationSnapshot> additionalImplementationsBuilder = ImmutableList.builder();
+    private final ImmutableSortedMap.Builder<String, HashCode> inputValueFingerprintsBuilder = ImmutableSortedMap.naturalOrder();
+    private final ImmutableSortedMap.Builder<String, CurrentFileCollectionFingerprint> inputFileFingerprintsBuilder = ImmutableSortedMap.naturalOrder();
+    private final ImmutableSortedMap.Builder<String, String> nonCacheableInputPropertiesBuilder = ImmutableSortedMap.naturalOrder();
+    private final ImmutableSortedSet.Builder<String> outputPropertiesBuilder = ImmutableSortedSet.naturalOrder();
+    private final ImmutableList.Builder<CachingDisabledReason> noCachingReasonsBuilder = ImmutableList.builder();
+
+    @Override
+    public void appendImplementation(ImplementationSnapshot implementation) {
+        this.implementation = implementation;
+        if (implementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                CachingDisabledReasonCatwgory.NON_CACHEABLE_IMPLEMENTATION,
+                "Implementation type " + implementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public void appendAdditionalImplementation(ImplementationSnapshot additionalImplementation) {
+        this.additionalImplementationsBuilder.add(additionalImplementation);
+        if (additionalImplementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                CachingDisabledReasonCatwgory.NON_CACHEABLE_ADDITIONAL_IMPLEMENTATION,
+                "Additional implementation type " + additionalImplementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public void appendInputValueFingerprint(String propertyName, ValueSnapshot fingerprint) {
+        Hasher hasher = Hashing.newHasher();
+        fingerprint.appendToHasher(hasher);
+        if (hasher.isValid()) {
+            HashCode hash = hasher.hash();
+            recordInputValueFingerprint(propertyName, hash);
+        } else {
+            markInputValuePropertyNotCacheable(propertyName, hasher.getInvalidReason());
+        }
+    }
+
+    protected void recordInputValueFingerprint(String propertyName, HashCode fingerprint) {
+        inputValueFingerprintsBuilder.put(propertyName, fingerprint);
+    }
+
+    protected void markInputValuePropertyNotCacheable(String propertyName, String nonCacheableReason) {
+        nonCacheableInputPropertiesBuilder.put(propertyName, nonCacheableReason);
+    }
+
+    @Override
+    public void appendInputFilesPropertyFingerprints(String propertyName, CurrentFileCollectionFingerprint fingerprints) {
```

Comment:
Why do we copy the whole immutable sorted map and create a new map instead of passing in the whole map at once?

Reply:
Good point, we could definitely do that.

Reply:
Done.